File: OGLEngine/BasicCamera.m
//
//  BasicCamera.m
//  jj
//
//  Created by Maciej Chmielewski on 14.02.2016.
//  Copyright © 2016 AppUnite. All rights reserved.
//

#import "BasicCamera.h"

@interface BasicCamera ()

@property (nonatomic, assign) GLKMatrix4 projectionMatrix;

@end

@implementation BasicCamera

- (instancetype)initWithPosition:(GLKVector3)position orientation:(GLKVector3)orientation {
    self = [self init];
    if (self) {
        self.position = position;
        self.orientation = orientation;
    }
    return self;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        float aspect = fabs([UIScreen mainScreen].bounds.size.width / [UIScreen mainScreen].bounds.size.height);
        self.projectionMatrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(65.0f), aspect, 0.1f, 100.0f);
    }
    return self;
}

- (GLKVector3)cameraPosition {
    return self.position;
}

- (GLKMatrix4)viewMatrix {
    GLKVector3 orientation = self.orientation;
    GLKVector3 position = self.position;
    GLKMatrix4 viewProjectionMatrix = GLKMatrix4Identity;
    viewProjectionMatrix = GLKMatrix4Rotate(viewProjectionMatrix, orientation.x, 1, 0, 0);
    viewProjectionMatrix = GLKMatrix4Rotate(viewProjectionMatrix, orientation.y, 0, 1, 0);
    viewProjectionMatrix = GLKMatrix4Rotate(viewProjectionMatrix, orientation.z, 0, 0, 1);
    viewProjectionMatrix = GLKMatrix4Translate(viewProjectionMatrix, position.x, position.y, position.z);
    return viewProjectionMatrix;
}

@end

File: OGLEngine/BasicCamera.swift
//
//  BasicCamera.swift
//  OGLEngine
//
//  Created by Maciej Chmielewski on 27.03.2016.
//  Copyright © 2016 MaciejCh. All rights reserved.
//

import Foundation
import GLKit

class BasicCamera: Camera {
    
    var position: GLKVector3! = GLKVector3Make(0, 0, 0)
    var orientation: GLKVector3! = GLKVector3Make(0, 0, 0)
    var staticProjectionMatrix: GLKMatrix4! = GLKMatrix4Identity
    
    init(position: GLKVector3, orientation: GLKVector3) {
        self.position = position
        self.orientation = orientation
        let aspect: Float = fabs(Float(UIScreen.mainScreen().bounds.size.width) / Float(UIScreen.mainScreen().bounds.size.height))
        self.staticProjectionMatrix = GLKMatrix4MakePerspective(GLKMathDegreesToRadians(65.0), aspect, 0.1, 100.0)
    }
    
    
    final func cameraPosition() -> GLKVector3 {
        return GLKVector3Make(-self.position.x, -self.position.y, -self.position.z)
    }
    
    func projectionMatrix() -> GLKMatrix4 {
        return self.staticProjectionMatrix
    }
    
    func viewMatrix() -> GLKMatrix4 {
        return transformatrionMatrix(self.position, orientation: self.orientation)
    }
    
    func viewProjectionMatrix() -> GLKMatrix4 {
        return self.viewMatrix() * self.projectionMatrix()
    }
}